### PR TITLE
[.Net] Add AutoGen.Mistral package

### PR DESCRIPTION
**Pull request by LittleLittleCloud**
## Why are these changes needed?
This PR is the first step to support Mistral in AutoGen.Net by adding the following functionality
- Add `AutoGen.Mistral` and `AutoGen.Mistral.Tests` project
- in `AutoGen.Mistral`, Add `MistralClient` which is in response in talking to mistral api. The `MistralClient` supports `json` mode and `function call`
- Streaming support in `MistralClient`
- Add `MistralClientAgent` which connects Mistral to AutoGen agent platform

> [!Note]
> `MistralClient` is just a utility class and shouldn't be considered as part of AutoGen-supported API. However I feel like it's still better to make it `public` in case it can be helpful in some situation..

## What remains to be done
~- Streaming support in `MistralClient`~
~- Add `MistralClientAgent` which connects Mistral to AutoGen agent platform~

## Related issue number
#2302 

## Checks

- [ ] I've included any doc changes needed for https://microsoft.github.io/autogen/. See https://microsoft.github.io/autogen/docs/Contribute#documentation to build and test documentation locally.
- [ ] I've added tests (if relevant) corresponding to the changes introduced in this PR.
- [ ] I've made sure all auto checks have passed.

### CODE REVIEW

**Review by ekzhu — COMMENTED** [review]
Do we want to implement our own client for Mistral API or using something off-the-shelf like https://github.com/tghamm/Mistral.SDK

**Comment by LittleLittleCloud** [comment]
> Do we want to implement our own client for Mistral API or using something off-the-shelf like https://github.com/tghamm/Mistral.SDK

I looked into all existing open source dotnet mistral sdks on GitHub (including mistral.sdk) and none of them satisfied our needs. For example, mistral.sdk doesn’t support function call. However in the future If mistral releases an official dotnet sdk we can migrate to using that package.

File: dotnet/src/AutoGen.Mistral/Agent/MistralClientAgent.cs [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+
+    private IEnumerable<ChatMessage> BuildChatHistory(IEnumerable<IMessage> messages)
+    {
+        var history = messages.Select(m => m switch
+        {
+            IMessage<ChatMessage> chatMessage => chatMessage.Content,
+            _ => throw new ArgumentException("Invalid message type")
+        });
+
+        // if there's no system message in the history, add one to the beginning
```
Comment (ekzhu) [inline dotnet/src/AutoGen.Mistral/Agent/MistralClientAgent.cs:98]: Is this a little unexpected for users if both _systemMessage and message in the chat history is set? I have encountered scenarios where task-specific system message instructions are needed. e.g., in a sequence of different chats with the same agent, each chat has different instructions.

File: dotnet/src/AutoGen.Mistral/Agent/MistralClientAgent.cs [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+            chatRequest.Tools = options.Functions.Select(f => new FunctionTool(f.ToMistralFunctionDefinition())).ToList();
+            chatRequest.ToolChoice = _toolChoice ?? ToolChoiceEnum.Auto;
+        }
+
+        return chatRequest;
+    }
+
+    private IEnumerable<ChatMessage> BuildChatHistory(IEnumerable<IMessage> messages)
+    {
+        var history = messages.Select(m => m switch
```
Comment (ekzhu) [inline dotnet/src/AutoGen.Mistral/Agent/MistralClientAgent.cs:92]: Is this a little unexpected for users if both _systemMessage is set and system messages exist in the chat history?  I have encountered scenarios where task-specific system message instructions are needed. e.g., in a sequence of different chats with the same agent, each chat has different instructions.

Maybe the more intuitive behavior is to have both _systemMessage and the input system messages in the history.

File: dotnet/src/AutoGen.Mistral/Middleware/MistralChatMessageConnector.cs
Comment (ekzhu) [inline dotnet/src/AutoGen.Mistral/Middleware/MistralChatMessageConnector.cs:1]: This connector middleware is not tested?

File: dotnet/src/AutoGen.Mistral/Agent/MistralClientAgent.cs (hunk 2 above)
Comment (ekzhu) [reply dotnet/src/AutoGen.Mistral/Agent/MistralClientAgent.cs:92]: It is possible to have multiple system messages, so some usage cases will take advantage of this. For now, we can make the behavior consistent with existing agent classes. We can make changes in the future the usage cases for .NET arises.

File: dotnet/src/AutoGen.Mistral/Middleware/MistralChatMessageConnector.cs
Comment (LittleLittleCloud) [reply dotnet/src/AutoGen.Mistral/Middleware/MistralChatMessageConnector.cs:1]: it's covered by the following tests
- MistralAgentFunctionCallMessageTest
- MistralAgentFunctionCallMiddlewareMessageTest
- MistralAgentFunctionCallAutoInvokeTestAsync
- MistralAgentFunctionCallTestAsync
- MistralAgentChatCompletionTestAsync

**Review by ekzhu — APPROVED** [review]

**Pull request merged**